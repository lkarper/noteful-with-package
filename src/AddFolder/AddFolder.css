.add-folder-form {
    display: flex;
    flex-direction: column;
}

.add-folder-form > * {
    margin: 5px;
}
