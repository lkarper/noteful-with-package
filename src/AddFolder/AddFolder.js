import React, { Component } from 'react';
import PropTypes from 'prop-types';
import NotesContext from '../NotesContext/NotesContext';
import ValidationError from '../ValidationError/ValidationError';
import config from '../config';
import './AddFolder.css';

class AddFolder extends Component {

    static contextType = NotesContext;

    static defaultProps = {
        history: {
            push: () => {},
        }
    }

    state = {
        folder: {
            name: '',
            touched: false,
        },
        error: null,
    }

    handleNewFolder = (event, cb) => {
        event.preventDefault();
        const newFolder = {
            name: this.state.folder.name,
        };

        fetch(`${config.API_URL}/folders`, {
            method: 'POST',
            body: JSON.stringify(newFolder),
            headers: {
                'content-type': 'application/json',
                'Authorization': `Bearer ${config.API_KEY}`
            },
        })
            .then(response => {
                if (response.ok) {
                    return response.json();
                }
                throw new Error(response.message);
            })
            .then(data => {
                cb(data);
                this.props.history.push(`/folder/${data.id}`);
            })
            .catch(error => {
                console.log(error);
                this.setState({ error: error.message });
            });
    }

    updateFolderName = (name) => {
        this.setState({
            folder: {
                name,
                touched: true,
            },
        });
    }

    validateFolderName = () => {
        const name = this.state.folder.name.trim();
        if (name.length === 0) {
            return 'Name is required and must contain at least one character.';
        }
    }

    render() {
        const { error } = this.state;
        const errorHTML = (
            <div className="folder-error">
                <h2>Looks like something went wrong: {error}.</h2>
                <p>A new folder could not be created at this time.  Please try again later.</p>
            </div>
        );
        const nameError = this.validateFolderName();

        return (
            <div>
                <form className="add-folder-form" onSubmit={e => this.handleNewFolder(e, this.context.addFolder)}>
                    <label htmlFor="folder-name">Enter a name for a new folder:</label>
                    <input 
                        type="text" 
                        id="folder-name" 
                        name="folder-name" 
                        onChange={e => this.updateFolderName(e.currentTarget.value)}
                        aria-required="true"
                        aria-describedby="aria-description"
                        aria-invalid={nameError ? true : false} 
                        required 
                    />
                    <span id="aria-description">You must enter a name for your folder.</span>
                    {this.state.folder.touched && <ValidationError message={nameError} />}
                    <button 
                        type="submit"
                        disabled={nameError ? true : false}
                        className="add-folder-button"
                    >
                        Create Folder
                    </button>
                </form>
                {error ?  errorHTML : ''}
            </div>
        );
    }
}

AddFolder.propTypes = {
    history: PropTypes.object.isRequired,
}

export default AddFolder;