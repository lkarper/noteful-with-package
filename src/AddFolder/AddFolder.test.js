import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import AddFolder from './AddFolder';

describe.skip('AddFolder component', () => {

    it('renders the UI without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(<AddFolder />, div);
        ReactDOM.unmountComponentAtNode(div);
    });

    it('renders the UI as expected', () => {
        const wrapper = shallow(<AddFolder />);
        expect(toJSON(wrapper)).toMatchSnapshot();
    });

    it('renders an error message when the input contains no characters', () => {
        const wrapper = shallow(<AddFolder />);
        wrapper.find('input').simulate('change', { currentTarget: {name: 'folder-name', value: ' '} });
        expect(toJSON(wrapper)).toMatchSnapshot();
    });

    it('disables submit button when input is blank', () => {
        const wrapper = shallow(<AddFolder />);
        wrapper.find('button').simulate('click');
        expect(toJSON(wrapper)).toMatchSnapshot();
    });

    it('enables submit button when valid folder name is entered', () => {
        const wrapper = shallow(<AddFolder />);
        wrapper.find('input').simulate('change', { currentTarget: {name: 'folder-name', value: 'valid folder name'} });
        expect(toJSON(wrapper)).toMatchSnapshot();
    });

});