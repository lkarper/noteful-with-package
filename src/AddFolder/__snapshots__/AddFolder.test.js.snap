// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddFolder component disables submit button when input is blank 1`] = `
<div>
  <form
    className="add-folder-form"
    onSubmit={[Function]}
  >
    <label
      htmlFor="folder-name"
    >
      Enter a name for a new folder:
    </label>
    <input
      aria-describedby="aria-description"
      aria-invalid={true}
      aria-required="true"
      id="folder-name"
      name="folder-name"
      onChange={[Function]}
      required={true}
      type="text"
    />
    <span
      id="aria-description"
    >
      You must enter a name for your folder.
    </span>
    <button
      className="add-folder-button"
      disabled={true}
      type="submit"
    >
      Create Folder
    </button>
  </form>
</div>
`;

exports[`AddFolder component enables submit button when valid folder name is entered 1`] = `
<div>
  <form
    className="add-folder-form"
    onSubmit={[Function]}
  >
    <label
      htmlFor="folder-name"
    >
      Enter a name for a new folder:
    </label>
    <input
      aria-describedby="aria-description"
      aria-invalid={false}
      aria-required="true"
      id="folder-name"
      name="folder-name"
      onChange={[Function]}
      required={true}
      type="text"
    />
    <span
      id="aria-description"
    >
      You must enter a name for your folder.
    </span>
    <ValidationError
      message=""
    />
    <button
      className="add-folder-button"
      disabled={false}
      type="submit"
    >
      Create Folder
    </button>
  </form>
</div>
`;

exports[`AddFolder component renders an error message when the input contains no characters 1`] = `
<div>
  <form
    className="add-folder-form"
    onSubmit={[Function]}
  >
    <label
      htmlFor="folder-name"
    >
      Enter a name for a new folder:
    </label>
    <input
      aria-describedby="aria-description"
      aria-invalid={true}
      aria-required="true"
      id="folder-name"
      name="folder-name"
      onChange={[Function]}
      required={true}
      type="text"
    />
    <span
      id="aria-description"
    >
      You must enter a name for your folder.
    </span>
    <ValidationError
      message="Name is required and must contain at least one character."
    />
    <button
      className="add-folder-button"
      disabled={true}
      type="submit"
    >
      Create Folder
    </button>
  </form>
</div>
`;

exports[`AddFolder component renders the UI as expected 1`] = `
<div>
  <form
    className="add-folder-form"
    onSubmit={[Function]}
  >
    <label
      htmlFor="folder-name"
    >
      Enter a name for a new folder:
    </label>
    <input
      aria-describedby="aria-description"
      aria-invalid={true}
      aria-required="true"
      id="folder-name"
      name="folder-name"
      onChange={[Function]}
      required={true}
      type="text"
    />
    <span
      id="aria-description"
    >
      You must enter a name for your folder.
    </span>
    <button
      className="add-folder-button"
      disabled={true}
      type="submit"
    >
      Create Folder
    </button>
  </form>
</div>
`;
